### PR TITLE
ILU Iter trisolve impl

### PR DESCRIPTION
**Pull request by liruipeng**
Copied from #711 

> Jacobi iterative solver for ILU smoother/preconditioner implemented on CPU/GPU currently for Nvidia only since we don't have AMD ILU factorization yet. This will occur in the next PR. This works with ILU0, ILU(k), and ILUT. For Schur compliment-based solves, we force tri_solve==1 during the setup and not allow iterative triangular solve during the solve phase. This will be implemented in a subsequent PR. When iterative tri solve is chosen for ILU0, ILU(k), an ILUT, one doesn't need to call solve analysis for the L and U factors. This yields a nontrivial performance gain.

- [x] tux
- [x] lassen
- [x] tioga
- [x] jlse

### CODE REVIEW

**Comment by PaulMullowney** [comment]
Any update on this PR? I have still have 1-2 more PRs after this to get the AMD device support working in the ILU classes.

**Comment by PaulMullowney** [comment]
Another ping. What's going on with this PR?

**Comment by oseikuffuor1** [comment]
> Another ping. What's going on with this PR?

Thanks for the reminder @PaulMullowney. We will finalize this PR soon.

File: src/utilities/device_utils.c [hunk 1]
```diff
@@ -2193,6 +2193,78 @@ hypreDevice_DiagScaleVector2( HYPRE_Int       n,
...
+   if (i < n)
+   {
+     z[i] = x[i] + alpha * y[i];
+   }
+}
+
+/*
+ */
+HYPRE_Int
+hypreDevice_zeqxmy(HYPRE_Int n, HYPRE_Complex *x, HYPRE_Complex alpha, HYPRE_Complex *y, HYPRE_Complex *z)
```
Comment (liruipeng) [inline src/utilities/device_utils.c]: duplicates `hypreDevice_Axpyn`?

File: src/utilities/device_utils.c [hunk 2]
```diff
@@ -2193,6 +2193,78 @@ hypreDevice_DiagScaleVector2( HYPRE_Int       n,
...
+   dim3 gDim = hypre_GetDefaultDeviceGridDimension(n, "thread", bDim);
+
+   HYPRE_GPU_LAUNCH( hypreGPUKernel_zeqxmy, gDim, bDim, n, x, alpha, y, z);
+
+   return hypre_error_flag;
+}
+
+
+/*****************************************************************
+ z[i] = (x[i] + alpha*y[i])/d[i]
```
Comment (liruipeng) [inline src/utilities/device_utils.c]: fix the comment `/d[i]`

File: src/utilities/device_utils.c (hunk 1 above)
Comment (PaulMullowney) [reply src/utilities/device_utils.c]: Sure looks the same to me. I'll fix.

File: src/utilities/device_utils.c (hunk 2 above)
Comment (PaulMullowney) [reply src/utilities/device_utils.c]: I'll fix.

**Review by oseikuffuor1 — APPROVED** [review]

**Pull request merged**